MdeModulePkg/Core/Dxe/Misc/MemoryAttributesTable.c

```diff
@@ -181,10 +181,10 @@ InstallMemoryAttributesTable (
   //
   MemoryAttributesTable = AllocatePool (sizeof (EFI_MEMORY_ATTRIBUTES_TABLE) + DescriptorSize * RuntimeEntryCount);
   ASSERT (MemoryAttributesTable != NULL);
-  MemoryAttributesTable->Version         = EFI_MEMORY_ATTRIBUTES_TABLE_VERSION;
+  MemoryAttributesTable->Version         = PcdGet32 (PcdMemoryAttributesTableVersion);
   MemoryAttributesTable->NumberOfEntries = RuntimeEntryCount;
   MemoryAttributesTable->DescriptorSize  = (UINT32)DescriptorSize;
-  if (gMemoryAttributesTableForwardCfi) {
+  if ((MemoryAttributesTable->Version > 0x01) && gMemoryAttributesTableForwardCfi) {

```

Comment:
What is the behavior of pre 6.2 Linux if Flags has a non zero value or Version >= 2?

Reply:
Output from dmesg of kernel 5.15:

```
[    0.000000] efi: EFI v2.70 by EDK II
[    0.000000] efi: SMBIOS=0x46d7a000 SMBIOS 3.0=0x46d78000 ACPI=0x46db6000 ACPI 2.0=0x46db6014 MEMATTR=0x438d2418 
[    0.000000] efi: memattr: Unexpected EFI Memory Attributes table version 2
```

Looking at the kernel sources, I will assume that it does not use MAT => its contents are not be used to apply protections to EFI runtime code and data

https://github.com/torvalds/linux/blob/v5.15/drivers/firmware/efi/memattr.c

Reply:
I am trying to understand why the PCD is needed.  Default of PCD is same as #define.  I do not see any functional changes from adding the PCD.  When would a platform use a value other than 2?

Reply:
You are right that the default behavior of PCD is the same as #define. However, with PCD we can easily override it in the DSC file of the platform, which we cannot do with #define. Code without my patch will support only EFI MAT v2. With this patch, you can choose whether to generate v1 or v2.

In v1, there was no `EFI_MEMORY_ATTRIBUTES_FLAGS_RT_FORWARD_CONTROL_FLOW_GUARD` flag. That's why we need such a condition:

> When would a platform use a value other than 2?

The EFI Memory Attributes table is defined by the UEFI Specification. The latest v2.10 specification refers to version 2. So there is a possibility that, at some point, there might also be a version 3 (since there was a v1).

Reply:
Hi! No response so far. Can I mark this one as resolved?

Reply:
If there is not use case yet, then I do not want a PCD added.  We try to minimize the number of PCDs because they are a list of items every platform integrator needs to review and verify.

Reply:
But there's a use case. You can produce MATv1 or MATv2. That's a cherry-pick from our corporate firmware. We need to support booting of some old 5.15 Linux kernels*.

\* But can we really call them old? According to https://www.kernel.org/category/releases.html EOL will be in December, 2026